### DISCUSSION TITLE
Leveraging AssetMapper to create an SPA/PWA

### DISCUSSION THREAD

**Discussion by tacman** [symfony/symfony#52832] [Q&A]
Now that the awkward build step is gone from deploying my Symfony apps (I love AssetMapper), I'm wondering the new tools can be leveraged to build a Progressive Web App.

I imagine that the best approach would be to first convert the app to an SPA (Single Page App).  My app gets most of its data from ApiPlatform, so it shouldn't be too hard to replace twig rendering calls with stimulus.  Or even fetch the HTML from turbo.  

I've seen some demo projects with React, and the Symfony book recommends preact and webpack (https://symfony.com/doc/6.2/the-fast-track/en/27-spa.html)  In particular, I guess that's my question: in light of AssetMapper, what's the best approach with Symfony 6.4 / 7?  

A tutorial or bundle for this would be awesome. This is interesting: https://www.twilio.com/blog/building-a-single-page-application-with-symfony-php-and-react, but uses webpack.

I realize that AssetMapper  isn't specifically the solution, but it feels so much easier that webpack that I'm thinking PWA's might not be as hard as I used to think.  Any guidance?  Thanks!

**Comment by tacman** [comment]
I found [this article,](https://alicia-paz.medium.com/make-your-rails-app-work-offline-part-1-pwa-setup-3abff8666194) and was hoping to find something like it but for Symfony instead of Rails.

The key line that jumped out is that it also uses importmap() 
   
> The Rails version is not as important as the fact that I’m not using a bundler to compile my assets.

And it uses Stimulus.  So it's got very similar components to a modern Symfony app.

**Comment by petrut-darius** [comment]
Hello how did you about this? I am new to symfony and I want to build an SPA app, but I don't know what to chose asset mapper or webpack, including the fact that I need it for my portfolio, cause I am looking for a php junior role

**Reply by tacman** [reply]
start here: https://github.com/Spomky-Labs/pwa-bundle  then grab some demos and play with it.  it's the right answer for using AssetMapper.

**Reply by petrut-darius** [reply]
Thank you.
